blog/_src/posts/2016-07-11-tutorial-racket-ffi-part-3.scrbl

```diff
@@ -0,0 +1,504 @@
+
+First, we'll start by defining a struct type for the Racket representation
+of Cairo paths:
+
+@examples[#:eval ev #:label #f
+(struct cairo-path (ptr)
+  #:property prop:sequence
+  (λ (p) (in-cairo-path p)))
+]
+
+The representation will store one field @racket[ptr] which, as the name
+suggests, will store a pointer value. We'll see what to do with this
+pointer later.
+
+This definition uses a @reftech{structure type property} to make instances of
+@racket[cairo-path] automatically work as sequences. This means that you
+can iterate over them with a @racket[for] loop or apply @racket[sequence-ref]
+on them. The property takes a function that takes an instance of the struct
+type itself (here @racket[p]) and that returns a sequence.
+
+We'll later define the @racket[in-cairo-path] function that will actually
+construct the relevant sequence for us. For now, let's see how to construct
+the C type given this struct type:
+
+@examples[#:eval ev #:label #f
+(define _cairo_path_t
+  (let ()
+    (code:comment "Extract pointer out of representation")
+    (define (racket->c rkt)
+      (cairo-path-ptr rkt))
+    (code:comment "Just apply the Racket constructor")
+    (define (c->racket cobj)
+      (cairo-path cobj))
+    (make-ctype _pointer
+                racket->c
+                c->racket)))
+]
+
+The base type for this @racket[_cairo_path_t] is a @racket[_pointer] type. Since
+the Cairo API returns pointers to new path values, it's hard to avoid using some
+kind of pointer type as the base type here.
+
+This definition right-hand-side defines the two conversion functions between
+Racket and C. Both are very simple because of how we've set up the representation.
+In the Racket to C case, we just extract the pointer field of the struct. In
+the other direction, we just stuff the pointer into a struct.
+
+The real work is done by the helper function that makes a
+@racket[cairo-path] instance work as a sequence.
+
+Starting top-down, let's look at the definition of @racket[in-cairo-path]:
+
+@examples[#:eval ev #:label #f
+(code:comment "Cairo-Path -> Sequence")
+(define (in-cairo-path path)
+  (define pp (cairo-path-ptr path))
+  (match-define
+    (list _ array-ptr len)
+    (ptr-ref pp _simple_cairo_path_t))
+  (make-do-sequence
+    (λ ()
+      (values (pos->element array-ptr)
+              (next-pos array-ptr)
+              0
+              (λ (pos) (< pos len))
+              #f #f))))
+]
+
+The first thing the function does is extract the pointer out of the
+representation, and then immediately calls @racket[ptr-ref] on it. This
+lets us manipulate the C path struct using the simple representation we
+defined in the first part of the blog post.
+
+Note: in case you're not very familiar with Racket pattern matching, the
+@racket[match-define] form is a very convenient form that lets you define
+potentially multiple variables using a pattern. The first argument clause
```

Comment:
Suggestion: "... the @racket[match-define] form lets you define potentially multiple variables using a pattern, similar to Haskell or OCaml's @racket[let] statement."